mountpoint-s3-fs/src/manifest/db.rs

```diff
@@ -0,0 +1,242 @@
+use rusqlite::{params_from_iter, Connection, Result, Row};
+use std::path::Path;
+use std::sync::{Arc, Mutex, MutexGuard};
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct DbEntry {
+    pub full_key: String, // Both files and directories don't have '/' in the end
+    pub etag: Option<String>,
+    pub size: Option<usize>,
+}
+
+impl DbEntry {
+    fn from_row(row: &Row) -> Result<Self> {
+        Ok(Self {
+            full_key: row.get(0)?,
+            etag: row.get(1)?,
+            size: row.get(2)?,
+        })
+    }
+
+    // Parent key without a trailing '/'
+    fn parent_key(&self) -> &str {
+        let key = self.full_key.trim_end_matches("/");
+        let last_component_len = key.rsplit("/").next().expect("expect at least one component").len();
+        if last_component_len == key.len() {
+            // root case is special, it doesn't contain trailing '/'
+            ""
+        } else {
+            &key[..key.len() - last_component_len - 1]
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(super) struct Db {
+    conn: Arc<Mutex<Connection>>,
+}
+
+impl Db {
+    pub(super) fn new(manifest_db_path: &Path) -> Result<Self> {
+        let conn = Connection::open(manifest_db_path)?;
+        let mode: String = conn.query_row("PRAGMA journal_mode=off", [], |row| row.get(0))?;
+        assert_eq!(&mode, "off");
+
+        Ok(Self {
+            conn: Arc::new(Mutex::new(conn)), // TODO: connection pool?
+        })
+    }
+
+    pub(super) fn create_table(&self) -> Result<()> {
+        let conn = self.conn.lock().expect("lock must succeed");
+        conn.execute(
+            "CREATE TABLE s3_objects (
+                id          INTEGER   PRIMARY KEY,
+                key         TEXT      NOT NULL,
+                parent_key  TEXT      NOT NULL,
+                etag        TEXT      NULL,
+                size        INTEGER   NULL
+            )",
+            (),
+        )?;
+
+        Ok(())
+    }
+
+    pub(super) fn create_index(&self) -> Result<()> {
+        let conn = self.conn.lock().expect("lock must succeed");
+        conn.execute("CREATE UNIQUE INDEX idx_key ON s3_objects (key)", ())?;
+
+        conn.execute("CREATE INDEX idx_parent_key ON s3_objects (parent_key, key)", ())?;
```

Comment:
Why? I don't think we ever filter using `parent_key` at all

Reply:
this PR just creates the DB, so all of the actual useful queries are not here, but here is how we [will](https://github.com/vladem/mountpoint-s3/pull/7/files#diff-c4e56cb844e7f3cd61b6fbcafc093275be050e5217317ebdfe15575d2f6f3bf5R73) do a `readdir` in future:
```
SELECT key, etag, size FROM s3_objects WHERE parent_key = ?1 ORDER BY key LIMIT ?2, ?3
```